### PR TITLE
Create new StreamPlatformExtension

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

This is an attempt to create a new StreamPlatformExtension, which should allow us to replace the TestStreams, StreamProcessorComposite, and StreamProcessorRule. This is something that would help us to continue with cleaning up the engine/stream processing code (and tests). There was a recent attempt to remove the event applies from the stream processing interface, https://github.com/camunda/zeebe/pull/9985 but this failed since several tests have high dependencies on how the engine and related stuff is built. They are highly coupled.

This is just a first step and I would like to validate with you @saig0 and @npepinpe, since I think you both are quite familiar with JUnit 5 extensions and our old rules. (Note: I'm a total junit 5 noob please bare with me.)

_What the PR does:_

 - Introduce a new class StreamPlatform (combination of [TestStreams](https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/util/TestStreams.java) and [StreamComposite](https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/util/StreamProcessingComposite.java))
 - Introduce a new extension to create the StreamPlatform and make this available for tests
 - Created an example test which is a copy of [StreamProcessorReplayModeTest.java](https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorReplayModeTest.java)

In the test class we can see how it would look like and what the stream platform could provide us. Instead of always registering mocked typed processors in streamProcessors tests we just have on mocked RecordProcessor which can be use for verifications. I think this would simplify our tests a lot, since it is currently a lot of boilter plate which is mostly for the tests not interesting at all.

Be aware that it currently only contains one test to show how it could look like and get early feedback. 
Please take a look and let me know what you think.

## Related issues

closes #

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/util/StreamPlatformExtension.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+  private Store getStore(final ExtensionContext context) {
+    return context.getStore(Namespace.create(getClass(), context.getUniqueId()));
+  }
+
+  public StreamProcessorTestContext lookupOrCreate(final ExtensionContext extensionContext) {
+    final var store = getStore(extensionContext);
+
+    return (StreamProcessorTestContext)
+        store.getOrComputeIfAbsent(FIELD_STATE, (key) -> new StreamProcessorTestContext());
```
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/util/StreamPlatformExtension.java:49]: Mostly copied from Peters https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/util/ZeebeStateExtension.java I figured out that it is useful for also closing resources

File: engine/src/test/java/io/camunda/zeebe/engine/util/StreamPlatform.java [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+  private final Map<String, LogContext> logContextMap = new HashMap<>();
+  private final Map<String, ProcessorContext> streamContextMap = new HashMap<>();
+  private boolean snapshotWasTaken = false;
+  private final StreamProcessorMode streamProcessorMode = StreamProcessorMode.PROCESSING;
+  private RecordProcessor recordProcessor;
+
+  private final WriteActor writeActor = new WriteActor();
+  private final ZeebeDbFactory zeebeDbFactory;
+
+  public StreamPlatform(
```
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/util/StreamPlatform.java:87]: Most parts are from TestStream and StreamProcessingComposite, I tried to clean up it already a bit

**Comment by npepinpe** [comment]
To clarify, what's the goal for this PR? If I understood correctly, you're blocked because the test classes are too tightly coupled with the engine API? I'm just trying to figure out what kind of feedback do you want :smile:

**Comment by Zelldon** [comment]
> To clarify, what's the goal for this PR? If I understood correctly, you're blocked because the test classes are too tightly coupled with the engine API? I'm just trying to figure out what kind of feedback do you want smile

Since I have zero experience with writing extensions I would like to have first general feedback: This looks like a valid extension or it is crap. 

Furthemore, I would like to know whether you think this is something we want to use to test our StreamProcessor. I would like that we add some point test the stream processor separately from the engine, with just a mocked or dummy RecordProcessor, similar like what is done in the PR. To just validate the API and its guarantees, but without coupling to the real engine processing logic. 

Blocked is a hard word I have also other things I can do but yes it would help to make progress in this part.

**Comment by Zelldon** [comment]
I see it as a first step in migrating our existing tests, so if you are ok with it and we merge it we can slowly migrate the other tests and delete at some point the old ones and hopefully at some point the old Rules. Of course, we can always iterate over the extension and StreamPlatform class (and we potentially also have to).

**Review by saig0 — APPROVED** [review]
@Zelldon the extension looks good so far :+1: 

I think that we can clean it up here and there. Especially, the `StreamPlatform` looks very similar to our existing classes. But I'm fine with improving it over time when we migrate more tests and know what we really need.

**Comment by Zelldon** [comment]
Thanks @saig0 for your review :) Yes makes sense since it is a copy/combination of them. As I said I also see that as a first step and we can improve but I think this helps us already to get the stone rolling :)

**Pull request merged**